src/cache.rs

```diff
@@ -226,33 +307,104 @@ impl<T: Hash + PartialEq> Constraint<T> {
     /// Enter a constraint for a call to an immutable function.
     #[inline]
     pub fn push(&self, args: T, ret: u128, mutable: bool) {
-        let both = hash(&(&args, ret));
-        self.push_inner(Call { args, ret, both, mutable });
+        let args_hash = hash(&args);
+        let both = hash(&(args_hash, ret));
+        self.0.write().push_inner(Cow::Owned(Call {
+            args,
+            args_hash,
+            ret,
+            both,
+            mutable,
+        }));
+    }
+
+    /// Whether the method satisfies as all input-output pairs.
+    #[inline]
+    pub fn validate<F>(&self, mut f: F) -> bool
+    where
+        F: FnMut(&T) -> u128,
+    {
+        self.0.read().calls.iter().all(|entry| f(&entry.args) == entry.ret)
+    }
+
+    /// Whether the method satisfies as all input-output pairs.
+    #[inline]
+    pub fn validate_with_id<F>(&self, mut f: F, id: usize) -> bool
+    where
+        F: FnMut(&T) -> u128,
+    {
+        let accelerator = accelerator(id);
+        let inner = self.0.read();
+        if let Some(accelerator) = accelerator {
+            let mut map = accelerator.lock();
+            inner.calls.iter().all(|entry| {
+                *map.entry(entry.both).or_insert_with(|| f(&entry.args)) == entry.ret
+            })
+        } else {
+            inner.calls.iter().all(|entry| f(&entry.args) == entry.ret)
+        }
```

Comment:
This is pretty much dead code on the mutable constraint since mutable validation can't be accelerated. It would only be triggered when a tracked type with &mut self tracked methods is used through a `Tracked` container instead of `TrackedMut` (which practically doesn't happen). 

We still need this method for the trait impl, but it could just as well dispatch to the normal `validate` with a short comment why. Less code == better.